base/set.jl

```diff
@@ -178,16 +178,41 @@ function unique(f, C)
     out
 end
 
-# If A is not grouped, then we will need to keep track of all of the elements that we have
-# seen so far.
-function _unique!(A::AbstractVector)
-    seen = Set{eltype(A)}()
+"""
+    unique!(f, A::AbstractVector)
+
+Selects one value from `A` for each unique value produced by `f` applied to
+elements of `A` , then return the modified A.
+
+# Examples
+```jldoctest
+julia> unique!(x -> x^2, [1, -1, 3, -3, 4])
+3-element Array{Int64,1}:
+ 1
+ 3
+ 4
+
+julia> unique!(n -> n%3, [5, 1, 8, 9, 3, 4, 10, 7, 2, 6])
+3-element Array{Int64,1}:
+ 5
+ 1
+ 9
+
+julia> unique!(iseven, [2, 3, 5, 7, 9])
+2-element Array{Int64,1}:
+ 2
+ 3
+```
+"""
+function unique!(f, A::AbstractVector)
+    seen = Set()
```

Comment:
Unfortunately this won't be efficient because this means `Set{Any}()`.

Since relying on inference is *naughty*, what should probably happen here is a widening pattern. Start with a `Set{Union{}}()` and then use an (imagined) function `push_widen!` which will widen the element type. Someone else might know better, but I *think* the most efficient pattern is then to dispatch again to a function like this. This is, unfortunately, a little bit of work to implement, sorry.

You'll want an inner function like `_unique!(f, A, set, i)` where `set` is the current set and `i` is the current index  The outer `unique!` function is creates the empty `Set` and dispatches to the inner one.

i.e. something like

```julia
function unique!(f, A::AbstractArray)
    if length(A) <= 1
        return A
    end

    i = firstindex(A)
    t = @inbounds A[i]
    seen = Set{typeof(x)}()
    push!(seen, t)
    return _unique!(f, A, seen, 1, i+1)
end
```

and the other funciton approximately

```julia
function _unique!(f, A::AbstractArray, seen::Set, count::Integer, i::Integer)
    while i < lastindex(A)
        t = @inbounds A[i]
        if t ∉ seen
            count += 1
            A[count] = t
            if typeof(t) <: eltype(seen)
                push!(seen, t)
            else
                seen2 = convert(Set{promote_typejoin(eltype(seen), typeof(t))}, seen)
                push!(seen2, t)
                return _unique!(f, A, seen2, count, i+1)
            end
        end
        i += 1
    end
    return resize!(A, count)
end
```

I haven't tested the above *at all*, but I hope it demonstrates the pattern at least. The compiler will elide all the complexity in the standard case that `f(first(A))` is inferrable. Also note that I've assumed all `AbstractVector`s have `AbstractUnitRange` indices in the above to avoid the need to use `iterate`, but the other pattern is fine too.

Reply:
While I agree,
I also think that can be in a seperate PR.
Right now, the existing `unique(f, iter)` (no bang),
does the same thing.
And can be optimised in the same way.

This PR adds a naive implementation that can be improved later.

Reply:
I also think this could be implemented in a different PR as both with and without bang version of unique(f, iter) uses naive method.

Reply:
@andyferris I can try to implement the inner function _unique within this PR if this is what is needed to merge it but will need some help.

Reply:
Sorry - I didn’t mean to block this - iterating later is of course fine.

Reply:
@andyferris you might want to dismiss your review,
However that is done,
(possibly by rereviewing?)

Reply:
@andyferris Please re review this PR.